app/Models/DatabaseDAO.php

```diff
@@ -164,4 +163,168 @@ public function ensureCaseInsensitiveGuids() {
+				$catFrom = $catDAO; $catTo = $categoryDAOSQLite;
+				$feedFrom = $feedDAO; $feedTo = $feedDAOSQLite;
+				$entryFrom = $entryDAO; $entryTo = $entryDAOSQLite;
+				$tagFrom = $tagDAO; $tagTo = $tagDAOSQLite;
+				break;
+			case self::SQLITE_IMPORT:
+				$userFrom = $userDAOSQLite; $userTo = $userDAO;
+				$catFrom = $categoryDAOSQLite; $catTo = $catDAO;
+				$feedFrom = $feedDAOSQLite; $feedTo = $feedDAO;
+				$entryFrom = $entryDAOSQLite; $entryTo = $entryDAO;
+				$tagFrom = $tagDAOSQLite; $tagTo = $tagDAO;
+				break;
+		}
+
+		$idMaps = [];
+
+		if (defined('STDERR')) {
+			fwrite(STDERR, "Start SQL copy…\n");
+		}
+
+		$userTo->createUser();
+
+		$catTo->beginTransaction();
+		foreach ($catFrom->select() as $category) {
+			$cat = $catTo->searchByName($category['name']);	//Useful for the default category
+			if ($cat != null) {
+				$catId = $cat->id();
+			} else {
+				$catId = $catTo->addCategory($category);
+				if ($catId == false) {
+					$error .= ' Error during SQLite copy of categories!';
+					goto done;
+				}
+			}
+			$idMaps['c' . $category['id']] = $catId;
+		}
+		foreach ($feedFrom->select() as $feed) {
+			$feed['category'] = empty($idMaps['c' . $feed['category']]) ? FreshRSS_CategoryDAO::DEFAULTCATEGORYID : $idMaps['c' . $feed['category']];
+			$feedId = $feedTo->addFeed($feed);
+			if ($feedId == false) {
+				$error .= ' Error during SQLite copy of feeds!';
+				goto done;
+			}
+			$idMaps['f' . $feed['id']] = $feedId;
+		}
+		$catTo->commit();
+
+		$nbEntries = $entryFrom->count();
+		$n = 0;
+		$entryTo->beginTransaction();
+		foreach ($entryFrom->select() as $entry) {
+			$n++;
+			if (!empty($idMaps['f' . $entry['id_feed']])) {
+				$entry['id_feed'] = $idMaps['f' . $entry['id_feed']];
+				if (!$entryTo->addEntry($entry, false)) {
+					$error .= ' Error during SQLite copy of entries!';
+					goto done;
+				}
+			}
+			if ($n % 100 === 1 && defined('STDERR')) {	//Display progression
+				fwrite(STDERR, "\033[0G" . $n . '/' . $nbEntries);
+			}
+		}
+		if (defined('STDERR')) {
+			fwrite(STDERR, "\033[0G" . $n . '/' . $nbEntries . "\n");
+		}
+		$entryTo->commit();
+		$feedTo->updateCachedValues();
+
+		$idMaps = [];
+
+		$tagTo->beginTransaction();
+		foreach ($tagFrom->select() as $tag) {
+			$tagId = $tagTo->addTag($tag);
+			if ($tagId == false) {
+				$error .= ' Error during SQLite copy of tags!';
+				goto done;
+			}
+			$idMaps['t' . $tag['id']] = $tagId;
+		}
+		foreach ($tagFrom->selectEntryTag() as $entryTag) {
+			if (!empty($idMaps['t' . $entryTag['id_tag']])) {
+				$entryTag['id_tag'] = $idMaps['t' . $entryTag['id_tag']];
+				if (!$tagTo->tagEntry($entryTag['id_tag'], $entryTag['id_entry'])) {
+					$error .= ' Error during SQLite copy of entry-tags!';
+					goto done;
+				}
+			}
+		}
+		$tagTo->commit();
+
+	done:
+		if ($error != '') {
+			if (defined('STDERR')) {
+				fwrite(STDERR, $error . "\n");
+			}
+			Minz_Log::error($error);
+			return false;
+		} else {
+			return true;
+		}
+	}
```

Comment:
This was a big part and I have to admit I was very sceptical when I saw the size of the method and the different calls to `goto`. In the end, it is pretty straightforward to read so I give you some insights but feel free to not consider them :)

- I would rather create two distinct method `import_from_sqlite($filename, $clearFirst = false)` and `export_to_sqlite($filename)` to don't have to manipulate a `$mode` parameter and its related `switch`es
- it implies a bit of factoring so I would have created `import` methods for each model (for instance `import_category($daoFrom, $daoTo)`) which would return `false` if an error occurs
- your `goto`s are nothing more than function calls but with a hidden `$error` parameter (`goto done;` vs `return done($error);`). While I understand the fun of using a pretty old notation next to the `yield`, I'm not sure that we gain in readability.

Reply:
I have removed the goto (there has been so much bashing during one generation), but before it dies:
* In this case, it is indeed a bit like calling a function, but better, because more lightweight (function calls are more expensive), better scoped, and with the code located at a more natural place in the flow.
* In a language that supports nested functions (e.g. C#, JavaScript, Pascal), this is what I would have used, but PHP does not support them, so now there is one more function out of scope...